Lib/zipfile.py

```diff
@@ -1147,8 +1147,15 @@ def writable(self):
     def write(self, data):
         if self.closed:
             raise ValueError('I/O operation on closed file.')
-        nbytes = len(data)
+
+        # Accept any data that supports the buffer protocol
+        if isinstance(data, (bytes, bytearray)):
+            nbytes = len(data)
+        else:
+            data = memoryview(data)
```

Comment:
I think it's recommended to manually [release](https://docs.python.org/3/library/stdtypes.html#memoryview.release) a `memoryview`. Implementations other than CPython may use a garbage collector that doesn't immediately finalize unreferenced objects. Maybe it's simpler to always use a `memoryview`:

```python
        # Accept any data that supports the buffer protocol
        with memoryview(data) as data:
            nbytes = data.nbytes
            self._file_size += nbytes
            self._crc = crc32(data, self._crc)
            if self._compressor:
                data = self._compressor.compress(data)
                self._compress_size += len(data)
            self._fileobj.write(data)
            return nbytes
```

Reply:
Good point.
There are some sites in stdlib have this problem, maybe they can be solved in another issue together.
In addition, I suspect if other Python implementations don't release the underlying buffer of `memoryview`, the program will run abnormally.

For the most common `bytes`/`bytearray` objects, it's also slower, since checking type is faster than creating a `memoryview` a little.

Let @serhiy-storchaka decide.

Reply:
I ran a benchmark,  the first is faster, 0.54 sec vs 0.59 sec.
```python
from time import perf_counter
from pyzstd import *

data = b'abc'
REPEAT = 300000

t1 = perf_counter()
for _ in range(REPEAT):
    if isinstance(data, (bytes, bytearray)):
        nbytes = len(data)
    else:
        data = memoryview(data)
    compress(data)
t2 = perf_counter()
print(t2-t1)

t1 = perf_counter()
for _ in range(REPEAT):
    with memoryview(data) as mv:
        length = mv.nbytes
        compress(mv)
t2 = perf_counter()
print(t2-t1)
```

Reply:
@eryksun is right. But it may be simpler to merge the current solution and fix issues with releasing memoryviews in all places uniformly later.